components/webgpu/canvas_context.rs

```diff
@@ -0,0 +1,771 @@
+        (self.image_size.width * self.image_format.bytes_per_pixel() as u32)
+            .next_multiple_of(wgt::COPY_BYTES_PER_ROW_ALIGNMENT)
+    }
+}
+
+/// State of staging buffer
+#[derive(Debug)]
+enum StagingBufferState {
+    /// Initial state, buffer has yet to be created,
+    /// only it's id is reserved
+    Unassigned,
+    /// Buffer is allocated on GPUDevice and ready to be used immediately
+    Available(Buffer),
+    /// Buffer is running mapAsync
+    Mapping(Buffer),
+    /// Buffer is currently actively mapped
+    Mapped(MappedBuffer),
+}
+
+/// Staging buffer used for
+/// Texture to Buffer to CPU copying
+#[derive(Debug)]
+struct StagingBuffer {
+    global: Arc<Global>,
+    buffer_id: BufferId,
+    state: StagingBufferState,
+}
+
+// [`StagingBuffer`] only used for reading (never for writing)
+// so it is safe to share between threads.
+unsafe impl Sync for StagingBuffer {}
+
+impl StagingBuffer {
+    fn new(global: Arc<Global>, buffer_id: BufferId) -> Self {
+        Self {
+            global,
+            buffer_id,
+            state: StagingBufferState::Unassigned,
+        }
+    }
+
+    const fn is_mapped(&self) -> bool {
+        matches!(self.state, StagingBufferState::Mapped(..))
+    }
+
+    /// Return true if buffer can be used directly with provided config
+    /// without any additional work
+    fn is_available_with_config(&self, config: &ContextConfiguration) -> bool {
+        let StagingBufferState::Available(buffer) = &self.state else {
+            return false;
+        };
+        buffer.has_config(config)
+    }
+
+    /// Return true if buffer is not mapping or being mapped
+    const fn needs_assign(&self) -> bool {
+        matches!(
+            self.state,
+            StagingBufferState::Unassigned | StagingBufferState::Available(_)
+        )
+    }
+
+    /// Make buffer available by unmap/destroy and recreating it if needed
+    fn ensure_available(&mut self, config: &ContextConfiguration) -> Result<(), CreateBufferError> {
+        let recreate = match &self.state {
+            StagingBufferState::Unassigned => true,
+            StagingBufferState::Available(buffer) |
+            StagingBufferState::Mapping(buffer) |
+            StagingBufferState::Mapped(MappedBuffer { buffer, .. }) => {
+                if buffer.has_config(config) {
+                    let _ = self.global.buffer_unmap(self.buffer_id);
+                    false
+                } else {
+                    self.global.buffer_drop(self.buffer_id);
+                    true
+                }
+            },
+        };
+        if recreate {
+            let buffer_size = config.buffer_size();
+            let (_, error) = self.global.device_create_buffer(
+                config.device_id,
+                &BufferDescriptor {
+                    label: None,
+                    size: buffer_size,
+                    usage: wgt::BufferUsages::MAP_READ | wgt::BufferUsages::COPY_DST,
```

Comment:
Please import this directly unless there is a naming conflict with another type, in general.
```suggestion
                    usage: BufferUsages::MAP_READ | BufferUsages::COPY_DST,
```